### PR TITLE
Updates to group formation and coarse tiling to support flash attention example

### PR DESCRIPTION
**Pull request by marnold-ibm**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [x] cleanup

#### What this PR does:
This PR fixes limitations in coarse tiling to allow the flash attention example to run with two tiled dimensions:  `{"H": 4, "Lk": 2}.

Changes include
- Corner cases in loop unrolling and tiling to handle ops that don't have all the dimensions that are tiled (broadcast)
- Cleanup: Combine number of temp variables being stored on the op into `_dim_prop_info`
- Move `assign_dim_hints` pass out of temp_passes into propagate_named_dims.py because it is no longer temp
- Flash attention test case added
#### Which issue(s) this PR is related to:
https://github.com/torch-spyre/torch-spyre/issues/2150
#### Additional note:

full flash driver that passes, and used to generate pytest case
```
import math

import torch
from torch_spyre._inductor import config, spyre_hint
import torch_spyre._inductor.propagate_named_dims as pnd

declare_tensor_dim = pnd.declare_tensor_dim
name_tensor_dims = pnd.name_tensor_dims

torch.manual_seed(0xAFFE)

B, H, Lq, Lk, D = 1, 8, 256, 256, 64
block_size = 128

queries_t = torch.randn(B, H, Lq, D, dtype=torch.float16)
keys_t = torch.randn(B, H, Lk, D, dtype=torch.float16)
values_t = torch.randn(B, H, Lk, D, dtype=torch.float16)


def flash(queries, keys, values):
    scale = 1.0 / math.sqrt(math.sqrt(D))

    output = torch.zeros_like(queries)
    M = torch.full(
        (B, H, Lq), float("-inf"), device=queries.device, dtype=torch.float16
    )

    with spyre_hint(slices={"B": 1}):
        with spyre_hint(slices={"H": 4}):
            with spyre_hint(slices={"Lk": Lk // block_size}):
                keys_T = keys.transpose(-1, -2).contiguous()

                denominator = torch.zeros((B, H, Lq), device=queries.device, dtype=torch.float16)
                scores = torch.matmul(queries * scale, keys_T * scale)  # B, H, Lq, Lk
                scores = scores.transpose(-1, -2).contiguous()  # avoid stick reduction: B, H, Lk, Lq
                block_max = torch.amax(scores, dim=-2)  # B, H, Lq
                max_running = torch.maximum(M, block_max)  # B, H, Lq

                exp_scores = torch.exp(
                    scores - max_running.unsqueeze(-2)
                )  # B, H, Lk, Lq
                correction = torch.exp(M - max_running)  # B, H, Lq

                denominator = denominator * correction + exp_scores.sum(dim=-2)  # B, H, Lq
                output = (
                    output * correction.unsqueeze(-1)
                    + torch.matmul(exp_scores.transpose(-1, -2), values)  # B, H, Lq, D
                )
                M = max_running
    return output / denominator.unsqueeze(-1)


# CPU reference
ref = flash(queries_t, keys_t, values_t)

# Device run
queries_dev = queries_t.to("spyre")
keys_dev = keys_t.to("spyre")
values_dev = values_t.to("spyre")

declare_tensor_dim("B", B)
declare_tensor_dim("H", H)
declare_tensor_dim("Lq", Lq)
declare_tensor_dim("Lk", Lk)
declare_tensor_dim("D", D)

name_tensor_dims(queries_dev, ["B", "H", "Lq", "D"])
name_tensor_dims(keys_dev, ["B", "H", "Lk", "D"])
name_tensor_dims(values_dev, ["B", "H", "Lk", "D"])  

config.coarse_tiling = True

try:
    result = torch.compile(flash)(queries_dev, keys_dev, values_dev).cpu()
except Exception as e:
    import traceback
    print(f"compile/run failed (expected during group analysis):")
    traceback.print_exc()
    raise SystemExit(0)

print("ref:   ", ref.flatten()[:8])
print("result:", result.flatten()[:8])
print("max abs diff:", torch.abs(ref - result).amax().item())

torch.testing.assert_close(
    result,
    ref,
    equal_nan=True,
    atol=1.0,
    rtol=0.1,
    msg=lambda msg: f"compiled spyre <-> cpu mismatch\n\n{msg}\n",
)
print("PASSED")

````

### CODE REVIEW

**Comment by marnold-ibm** [comment]
Deleted:  comment was in wrong PR

**Review by dgrove-oss — DISMISSED** [review]
lgtm

**Review by dgrove-oss — APPROVED** [review]
Approving. 

Code review found some existing issues and some small items, but they will be fixed in a followup PR  because the fix to the broadcast-only op and related pool allocation problem are significant and should be reviewed independent of the other cleanups and fixes in this PR.

Including review comments for the record:

--------


  Found 5 issues:

  1. Module-level docstring shows old 2-tuple nested spec format (score: 100)

  The module docstring example in coarse_tile.py shows [(K1, [0]), (K2, [1])], but the PR changes the spec to (hint_id, loop_count, tiled_dims) 3-tuples. The function-level Parameters was updated but the module example
  wasn't. Following the module example would produce wrong indexing in _stamp_group.

  → torch_spyre/_inductor/coarse_tile.py lines 42–57

  2. _log_op_inputs not updated after named_dims moved to _dim_prop_info (score: 100)

  _log_op_inputs uses getattr(buf, "named_dims", "?") but named_dims was migrated to buf._dim_prop_info.named_dims. Will silently log "?" for all input buffers, breaking the dim-propagation debug output.

  → torch_spyre/_inductor/propagate_named_dims.py lines 243–248

  3. New loop-invariant early-return in _propagate_tiled_op skips outside-consumer patching (score: 75)

  The new early-return for loop-invariant ops (all(not dims ...)) fires before _find_outside_consumers is evaluated. If such an op has outside consumers or is a graph output, they won't be patched to read from a full-size
  buffer. Also, the default [[]] means ops with loop_group_id but no loop_tiled_dims attribute are silently treated as loop-invariant.

  → torch_spyre/_inductor/coarse_tile.py (new block ~line 193–204)

  4. Broadcast-only ops silently dropped from coarse tiling groups (score: 75)

  The if spec: guard in hints_to_coarse_tile_groups discards groups where all ops have dim_index=None. Those ops get no loop_group_id while siblings in the same hint scope do — inconsistent stamping. This is directly
  relevant to flash attention accumulator ops (the PR's target use case), and the skipped test suggests correctness isn't established yet.

  → torch_spyre/_inductor/temp_passes.py lines ~944–964

  5. hinted_names[0] silently drops additional hinted names on the same loop variable (score: 75)

  In assign_dim_hints, when a loop var matches multiple hinted names, only hinted_names[0] is used to look up the split count. If two names map to the same loop var with different split counts, the second is silently
  ignored. Carried forward from a prior review note on a related PR.

  → torch_spyre/_inductor/propagate_named_dims.py lines ~701–720
